common/src/main/java/io/netty/util/concurrent/SingleThreadEventExecutor.java

```diff
@@ -979,81 +1093,115 @@ private boolean ensureThreadStarted(int oldState) {
     }
 
     private void doStartThread() {
-        assert thread == null;
         executor.execute(new Runnable() {
             @Override
             public void run() {
+                processingLock.lock();
+                assert thread == null;
                 thread = Thread.currentThread();
                 if (interrupted) {
                     thread.interrupt();
+                    interrupted = false;
                 }
-
                 boolean success = false;
                 updateLastExecutionTime();
                 try {
-                    SingleThreadEventExecutor.this.run();
-                    success = true;
-                } catch (Throwable t) {
-                    logger.warn("Unexpected exception from an event executor: ", t);
-                } finally {
                     for (;;) {
-                        int oldState = state;
-                        if (oldState >= ST_SHUTTING_DOWN || STATE_UPDATER.compareAndSet(
-                                SingleThreadEventExecutor.this, oldState, ST_SHUTTING_DOWN)) {
-                            break;
+                        SingleThreadEventExecutor.this.run();
+                        success = true;
+
+                        if (state == ST_SUSPENDING && !hasTasks() && nextScheduledTaskDeadlineNanos() == -1) {
+                            if (!STATE_UPDATER.compareAndSet(SingleThreadEventExecutor.this,
+                                    ST_SUSPENDING, ST_SUSPENDED)) {
+                                // Try again
+                                continue;
+                            }
                         }
+                        break;
                     }
 
-                    // Check if confirmShutdown() was called at the end of the loop.
-                    if (success && gracefulShutdownStartTime == 0) {
-                        if (logger.isErrorEnabled()) {
-                            logger.error("Buggy " + EventExecutor.class.getSimpleName() + " implementation; " +
-                                    SingleThreadEventExecutor.class.getSimpleName() + ".confirmShutdown() must " +
-                                    "be called before run() implementation terminates.");
+                } catch (Throwable t) {
+                    logger.warn("Unexpected exception from an event executor: ", t);
+                } finally {
```

Comment:
Without having dug through every line, the shutdown sequence now feels like it is really two different shutdown mixed together. Can they be separated?

Reply:
I think its fine